### PR TITLE
[0837] Migrate the PE allocations section open state

### PR DESCRIPTION
**Pull request by defong**
### Context
Migrate the PE allocations section open state

### Changes proposed in this pull request
Added repeat allocation flow
Added edit initial allocation flow
Added choose an org allocation flow

#### Providers with repeat allocations

image

https://teacher-training-api-pr-2631.london.cloudapps.digital/publish/organisations/D86/2022/allocations
https://teacher-training-api-pr-2631.london.cloudapps.digital/publish/organisations/E69/2022/allocations
https://teacher-training-api-pr-2631.london.cloudapps.digital/publish/organisations/2BA/2022/allocations
https://teacher-training-api-pr-2631.london.cloudapps.digital/publish/organisations/L51/2022/allocations
https://teacher-training-api-pr-2631.london.cloudapps.digital/publish/organisations/L24/2022/allocations
https://teacher-training-api-pr-2631.london.cloudapps.digital/publish/organisations/C55/2022/allocations
https://teacher-training-api-pr-2631.london.cloudapps.digital/publish/organisations/G12/2022/allocations
https://teacher-training-api-pr-2631.london.cloudapps.digital/publish/organisations/1S2/2022/allocations
https://teacher-training-api-pr-2631.london.cloudapps.digital/publish/organisations/1WU/2022/allocations
https://teacher-training-api-pr-2631.london.cloudapps.digital/publish/organisations/E65/2022/allocations
https://teacher-training-api-pr-2631.london.cloudapps.digital/publish/organisations/24P/2022/allocations
https://teacher-training-api-pr-2631.london.cloudapps.digital/publish/organisations/B52/2022/allocations
https://teacher-training-api-pr-2631.london.cloudapps.digital/publish/organisations/O66/2022/allocations
https://teacher-training-api-pr-2631.london.cloudapps.digital/publish/organisations/B71/2022/allocations
https://teacher-training-api-pr-2631.london.cloudapps.digital/publish/organisations/25G/2022/allocations
https://teacher-training-api-pr-2631.london.cloudapps.digital/publish/organisations/28E/2022/allocations

#### Provider without repeat allocations
image

https://teacher-training-api-pr-2631.london.cloudapps.digital/publish/organisations/1WH/2022/allocations
https://teacher-training-api-pr-2631.london.cloudapps.digital/publish/organisations/S25/2022/allocations
https://teacher-training-api-pr-2631.london.cloudapps.digital/publish/organisations/26U/2022/allocations
https://teacher-training-api-pr-2631.london.cloudapps.digital/publish/organisations/1SG/2022/allocations
https://teacher-training-api-pr-2631.london.cloudapps.digital/publish/organisations/1A9/2022/allocations
https://teacher-training-api-pr-2631.london.cloudapps.digital/publish/organisations/1SH/2022/allocations
https://teacher-training-api-pr-2631.london.cloudapps.digital/publish/organisations/2H8/2022/allocations
https://teacher-training-api-pr-2631.london.cloudapps.digital/publish/organisations/28T/2022/allocations
https://teacher-training-api-pr-2631.london.cloudapps.digital/publish/organisations/K60/2022/allocations
https://teacher-training-api-pr-2631.london.cloudapps.digital/publish/organisations/F82/2022/allocations
https://teacher-training-api-pr-2631.london.cloudapps.digital/publish/organisations/2A6/2022/allocations
https://teacher-training-api-pr-2631.london.cloudapps.digital/publish/organisations/2BB/2022/allocations
https://teacher-training-api-pr-2631.london.cloudapps.digital/publish/organisations/B60/2022/allocations
https://teacher-training-api-pr-2631.london.cloudapps.digital/publish/organisations/B31/2022/allocations
https://teacher-training-api-pr-2631.london.cloudapps.digital/publish/organisations/252/2022/allocations
https://teacher-training-api-pr-2631.london.cloudapps.digital/publish/organisations/N73/2022/allocations

#### Notes
image

### Checklist

- [x] Make sure all information from the Trello card is in here
- [x] Attach to Trello card
- [x] Rebased master
- [x] Cleaned commit history
- [x] Tested by running locally

### CODE REVIEW

**Comment by defong** [comment]
[Migrate the PE allocations section, open state](https://trello.com/c/5XCfNTs7/837-migrate-the-pe-allocations-section-open-state)

File: app/flows/initial_request_flow.rb [hunk 1]
```diff
@@ -0,0 +1,218 @@
...
+
+  def training_provider_search_service
+    API::V2::AccreditedProviderTrainingProvidersController::TrainingProviderSearch
+  end
+
+  def training_providers_with_fee_paying_pe_course
+    @training_providers_with_fee_paying_pe_course ||=
+
+      training_provider_search_service.new(
+        provider: provider, params: { filter: { subjects: PE_SUBJECT_CODE,
```
Comment (darokel) [inline app/flows/initial_request_flow.rb:129]: This is some dodgy indentation

File: spec/support/page_objects/publish/allocations/request/who_are_you_requesting_a_course_for_page.rb [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+
+          sections :radio_buttons, Sections::RadioButton, ".govuk-radios__item"
+
+          element :continue, ".govuk-button"
+
+          element :training_provider_search, "#training-provider-query-field"
+
+          def providers
+            @providers ||=
+              radio_buttons.reject do |x| x == find_an_organisation_not_listed_above end
```
Comment (darokel) [inline spec/support/page_objects/publish/allocations/request/who_are_you_requesting_a_course_for_page.rb:22]:
```suggestion
radio_buttons.reject { |x| x == find_an_organisation_not_listed_above }
```

File: spec/support/page_objects/publish/allocations/request/who_are_you_requesting_a_course_for_page.rb [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+          end
+
+          def find_an_organisation_not_listed_above
+            @find_an_organisation_not_listed_above ||= radio_buttons.find do |x| x.text == "Find an organisation not listed above" end
```
Comment (darokel) [inline spec/support/page_objects/publish/allocations/request/who_are_you_requesting_a_course_for_page.rb:26]:
```suggestion
@find_an_organisation_not_listed_above ||= radio_buttons.find { |x| x.text == "Find an organisation not listed above" }
```

File: app/controllers/publish/providers/allocations_controller.rb [hunk 4]
```diff
@@ -9,8 +9,97 @@ def index
         )
       end
 
+      def new_repeat_request
+        authorize(Allocation)
+        provider
```
Comment (darokel) [inline app/controllers/publish/providers/allocations_controller.rb:14]: is it worth cleaning up the instance_var (eg, `provider, `training_provider` loading into before_actions for the relevant action? Or did you want to keep it explicit when they're used/loaded

**Review by darokel — APPROVED** [review]
Nice work getting this over the line 💪

**Comment by darokel** [comment]
Something I just noticed..are the table widths supposed to span over the width? eg

Screenshot 2022-05-10 at 13 46 37

**Comment by defong** [comment]
> Something I just noticed..are the table widths supposed to span over the content width? eg
> 
> Screenshot 2022-05-10 at 13 46 37

It's per design, the text is two third width and the table is full width.

See
https://bat-design-history.netlify.app/publish-teacher-training-courses/request-pe-courses/#new-organisation-added

File: app/flows/initial_request_flow.rb (hunk 1 above)
Comment (defong) [reply app/flows/initial_request_flow.rb]: Fixed, we really should invest in our linter through.

File: spec/support/page_objects/publish/allocations/request/who_are_you_requesting_a_course_for_page.rb (hunk 2 above)
Comment (defong) [reply spec/support/page_objects/publish/allocations/request/who_are_you_requesting_a_course_for_page.rb]: Fixed, we really should invest in our linter through.

File: spec/support/page_objects/publish/allocations/request/who_are_you_requesting_a_course_for_page.rb (hunk 3 above)
Comment (defong) [reply spec/support/page_objects/publish/allocations/request/who_are_you_requesting_a_course_for_page.rb]: Fixed, we really should invest in our linter through.

File: app/controllers/publish/providers/allocations_controller.rb (hunk 4 above)
Comment (defong) [reply app/controllers/publish/providers/allocations_controller.rb:14]: https://github.com/DFE-Digital/teacher-training-api/blob/1a0a404b9772902f9460d26ffbd2642f4ee04b98/docs/adr/0004-before-action-variables.md

**Review by mbocc — APPROVED** [review]
Thanks for the massive work!!

**Pull request merged**